src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5790,149 +5792,212 @@ public static Task WhenAll(IEnumerable<Task> tasks)
+        /// where its exceptions will contain the aggregation of the set of unwrapped exceptions from each of the supplied tasks.
+        /// </para>
+        /// <para>
+        /// If none of the supplied tasks faulted but at least one of them was canceled, the returned task will end in the Canceled state.
+        /// </para>
+        /// <para>
+        /// If none of the tasks faulted and none of the tasks were canceled, the resulting task will end in the RanToCompletion state.
+        /// </para>
+        /// <para>
+        /// If the supplied span contains no tasks, the returned task will immediately transition to a RanToCompletion
+        /// state before it's returned to the caller.
+        /// </para>
+        /// </remarks>
+        /// <exception cref="System.ArgumentException">The <paramref name="tasks"/> array contained a null task.</exception>
+        internal static Task WhenAll(ReadOnlySpan<Task> tasks) => // TODO https://github.com/dotnet/runtime/issues/77873: Make this public.
+            tasks.Length != 0 ? new WhenAllPromise(tasks) : CompletedTask;
 
-        // A Task that gets completed when all of its constituent tasks complete.
-        // Completion logic will analyze the antecedents in order to choose completion status.
-        // This type allows us to replace this logic:
-        //      Task promise = new Task(...);
-        //      Action<Task> completionAction = delegate { <completion logic>};
-        //      TaskFactory.CommonCWAllLogic(tasksCopy).AddCompletionAction(completionAction);
-        //      return promise;
-        // which involves several allocations, with this logic:
-        //      return new WhenAllPromise(tasksCopy);
-        // which saves a couple of allocations and enables debugger notification specialization.
-        //
-        // Used in InternalWhenAll(Task[])
+        /// <summary>A Task that gets completed when all of its constituent tasks complete.</summary>
         private sealed class WhenAllPromise : Task, ITaskCompletionAction
         {
-            /// <summary>
-            /// Stores all of the constituent tasks.  Tasks clear themselves out of this
-            /// array as they complete, but only if they don't have their wait notification bit set.
-            /// </summary>
-            private readonly Task?[] m_tasks;
+            /// <summary>Either a single faulted/canceled task, or a list of faulted/canceled tasks.</summary>
+            private object? _failedOrCanceled;
             /// <summary>The number of tasks remaining to complete.</summary>
-            private int m_count;
+            private int _remainingToComplete;
 
-            internal WhenAllPromise(Task[] tasks)
+            internal WhenAllPromise(ReadOnlySpan<Task> tasks)
             {
-                Debug.Assert(tasks != null, "Expected a non-null task array");
```

Comment:
Out of curiosity, what's the reasoning behind removing this Debug Assert?

Reply:
`ReadOnlySpan<T>` can't be null.